pandas/core/arrays/arrow/array.py

```diff
@@ -127,12 +127,26 @@ def floordiv_compat(
         left: pa.ChunkedArray | pa.Array | pa.Scalar,
         right: pa.ChunkedArray | pa.Array | pa.Scalar,
     ) -> pa.ChunkedArray:
-        # Ensure int // int -> int mirroring Python/Numpy behavior
-        # as pc.floor(pc.divide_checked(int, int)) -> float
-        converted_left = cast_for_truediv(left, right)
-        result = pc.floor(pc.divide(converted_left, right))
-        if pa.types.is_integer(left.type) and pa.types.is_integer(right.type):
+        divided = pc.divide(left, right)
+        if pa.types.is_integer(divided.type):
+            # GH 56676, avoid storing intermediate calculating in floating point type.
+            has_remainder = pc.not_equal(pc.multiply(divided, right), left)
```

Comment:
IMO floordiv would be a useful compute kernel in arrow, especially since python and numpy both support it. I opened https://github.com/apache/arrow/issues/39386, if added would make this code much simpler (and probably perform better).